File: Day6.py
# class Solution(object):
#     def convert(s, numRows):
#         return s[::numRows * 2 - 2] + ''.join([l + s[numRows * 2 - i - 2::numRows * 2 - 2][k] if k < len(s[numRows * 2 - i - 2::numRows * 2 - 2]) else l for i in range(1, numRows-1) for k, l in enumerate(s[i::numRows * 2 - 2])]) + s[numRows - 1::numRows * 2 - 2] if numRows > 1 else s


#Second
# class Solution(object):
#     def convert(s, numRows):
#         if numRows == 1:
#             return s

#         ret = ""
#         ls = len(s)

#         curRow = 0
#         while curRow < numRows:
#             for i in range(curRow, ls, 2 * numRows - 2):
#                 ret += s[i]
#                 if curRow > 0 and curRow < numRows - 1 and i + 2 * (numRows - curRow) - 2 < ls:
#                     ret += s[i + 2 * (numRows - curRow) - 2]

#             curRow += 1
    
#         return ret


#Third Solution
class Solution(object):
    def convert(s, numRows):
        idx = [i for i in range(numRows)]+[i for i in range(-2,-1*numRows,-1)] 
        l, strs = len(idx), ['' for _ in range(numRows)]
        for i,c in enumerate(s):
            strs[idx[i%l]] += c
        return ''.join(strs)

    s = "PAYPALISHIRING"
    numRows = 5
    y = convert( s,numRows)
    print (y)

File: Day6.js
//Welcome To The Three Ways Series To solve a Problem::

/*Problem Statement:  
       Write the Function that will take a string and
       make this conversion given a number of rows:
       Ex:
       Input: s = "PAYPALISHIRING", numRows = 3
       Output: "PAHNAPLSIIGYIR"
*/

// 1way

// function convert(s, numRows) {
//   if (numRows === 1) return s;
//   const matrix = [];
//   for (let i = 0; i < numRows; i++) {
//     matrix[i] = [];
//   }
//   for (let i = 0, j = 0, dir = 1; i < s.length; i++, j += dir) {
//     matrix[j].push(s.charAt(i));
//     if (j === 0) dir = 1;
//     if (j === numRows - 1) dir = -1;
//   }
//   return matrix.reduce((res, row) => res + row.join(""), "");
// }

//2 way

// function convert(s, numRows) {
//   if (numRows === 1) {
//     return s;
//   }
//   const N = s.length;
//   const arr = [...Array(numRows)].map((r) => []);

//   for (let i = 0; i < N; i++) {
//     const pos = i % (2 * numRows - 2);
//     const ii = pos < numRows ? pos : 2 * numRows - 2 - pos;
//     arr[ii].push(s[i]);
//   }
//   return arr.map((r) => r.join("")).join("");
// }

// 3 way

function convert(s, numRows) {
  const zigzag = [...new Array(numRows).keys()];

  zigzag.push(...zigzag.slice(1, -1).reverse());

  const rows = new Array(numRows).fill("");
  [...s].forEach((c, i) => (rows[zigzag[i % zigzag.length]] += c));
  return rows.join("");
}

let s = "PAYPALISHIRING";
let numRows = 5;
console.log(convert(s, numRows));

numRows = 4;
console.log(convert(s, numRows));

numRows = 7;
console.log(convert(s, numRows));

/* Test case:
      1) 
         s = "PAYPALISHIRING" , numRows= 5 = PHASIYIRPLIGAN
      2)
         s = "PAYPALISHIRING" , numRows= 4 = PINALSIGYAHRPI
      3)
         s = "PAYPALISHIRING" , numRows= 7 = PNAIGYRPIAHLSI
 */
